src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java

```diff
@@ -18,870 +18,924 @@
+        if (prefix.endsWith("/"))   prefix = prefix.substring(0, prefix.length()-1);
+        if (suffix.startsWith("/")) suffix = suffix.substring(1, suffix.length());
+
+        return prefix + '/' + suffix;
+    }
+
+    public synchronized void terminateJenkinsSlave(String name) {
+        LOGGER.info("Terminating jenkins slave " + name);
+
+        TaskID taskId = TaskID.newBuilder().setValue(name).build();
+
+        if (results.containsKey(taskId)) {
+            LOGGER.info("Killing mesos task " + taskId);
+            driver.killTask(taskId);
+        } else {
+            // This is handling the situation that a slave was provisioned but it never
+            // got scheduled because of resource scarcity and jenkins later tries to remove
+            // the offline slave but since it was not scheduled we have to remove it from
+            // the request queue. The method has been also synchronized because there is a race
+            // between this removal request from jenkins and a resource getting freed up in mesos
+            // resulting in scheduling the slave and resulting in orphaned task/slave not monitored
+            // by Jenkins.
+
+            for(Request request : requests) {
+                if(request.request.slave.name.equals(name)) {
+                    LOGGER.info("Removing enqueued mesos task " + name);
+                    requests.remove(request);
+                    // Also signal the Thread of the MesosComputerLauncher.launch() to exit from latch.await()
+                    // Otherwise the Thread will stay in WAIT forever -> Leak!
+                    request.result.failed(request.request.slave);
+                    return;
+                }
             }
-            requests.remove(request);
-            processedRequests++;
-            break;
-          }
+
+            LOGGER.warning("Asked to kill unknown mesos task " + taskId);
+        }
+
+        // Since this task is now running, we should not start this task up again at a later point in time
+        finishedTasks.add(taskId);
+
+        if (mesosCloud.isOnDemandRegistration()) {
+            supervise();
         }
-      }
 
-      if (!taskCreated) {
-        driver.declineOffer(offer.getId());
-      }
     }
-    if (processedRequests > 0) {
-      if (!requests.isEmpty()) {
-        LOGGER.info("Created " + processedRequests + " tasks from " + offers.size() + " offers (" + requests.size() + " pending requests).");
-      } else {
-        LOGGER.info("Created " + processedRequests + " tasks from " + offers.size() + " offers.");
-      }
-    } else {
-      if (!requests.isEmpty()) {
-        LOGGER.info("Did not match any of the " + offers.size() + " offers (" + requests.size() + " pending requests)");
-      }
+
+    @Override
+    public void registered(SchedulerDriver driver, FrameworkID frameworkId, MasterInfo masterInfo) {
+        LOGGER.info("Framework registered! ID = " + frameworkId.getValue());
```

Comment:
Related to line 391: Wondering if Mesos would call this method with 'null' for frameworkId then, frameworkId.getValue() would throw a null pointer exception. It may need a null check ?